File: src/main/java/com/mmust/tokenizer/Tokenizer.java
package com.mmust.tokenizer;

import com.mmust.token.*;

import java.util.Iterator;

import static java.lang.String.valueOf;

public class Tokenizer implements Iterable<Token> {

	private char[] data;

	public Tokenizer(String expression) {
		this.data = expression.toCharArray();
	}

	private Token parseToken(char c) {
		Token token;
		if (c == '(') {
			token = Parenthesis.LEFT;
		} else if (c == ')') {
			token = Parenthesis.RIGHT;
		} else if (Character.isDigit(c)) {
			token = new NumToken(valueOf(c));
		} else if (c == '+') {
			token = new AdditionOpToken();
		} else if (c == '-') {
			token = new SubtractOpToken();
		} else if (c == '*') {
			token = new MultiplicationOp();
		} else if (c == '/') {
			token = new DivisionOpToken();
		} else {
			throw new IllegalArgumentException(String.format("Can't recognize token type, token = [%s]", c));
		}
		return token;
	}

	@Override
	public Iterator<Token> iterator() {
		return new Itr();
	}

	private class Itr implements Iterator<Token> {
		private int pos = 0;

		@Override
		public boolean hasNext() {
			return pos < data.length;
		}

		@Override
		public Token next() {
			return parseToken(data[pos++]);
		}
	}
}

File: src/main/kotlin/com/mmust/tokenizer/Tokenizer.kt
package com.mmust.tokenizer

import com.mmust.token.*

class Tokenizer(expression: String) {

    private val data = expression
    private var pos = 0

    fun parse(): List<Token> {

        val result = mutableListOf<Token>()

        do {
            val c = peek()
            val token = when {
                c == '(' -> LeftParenthesis
                c == ')' -> RightParenthesis
                Character.isDigit(c) -> extractDigit()
                c == '+' -> ArithmeticOperators.PLUS
                c == '-' -> ArithmeticOperators.MINUS
                c == '*' -> ArithmeticOperators.TIMES
                c == '/' -> ArithmeticOperators.DIV
                else -> throw IllegalArgumentException("Can't recognize token type, token = [$c]")
            }
            result.add(token)
            movePos()
        } while (hasNext())

        return result
    }

    private fun extractDigit(): Token {
        var r = peek().toString()

        while (oneMoreDigitPresent()) {
            movePos()
            r += peek()
        }

        return NumToken(r)
    }

    private fun peek() = data[pos]

    private fun hasNext() = pos < data.length

    private fun oneMoreDigitPresent() = pos + 1 < data.length && data[pos + 1].isDigit()

    private fun movePos(offset: Int = 1) {
        pos += offset
    }
}
